### ISSUE TITLE
Incorrect bounding boxes, same for all images

### ISSUE THREAD

**Issue by mjohnst** [thtrieu/darkflow#172]
Hello,

I am attempting to retrain one of the YOLO models with my own data and only 1 class.
I've tried both YOLO v1 and v2, but got a bit closer to something working in v1.

I started by copying `v1.1/tiny-yolo-4c.cfg` into my own cfg and changing `classes=1` and `output=539` 
`side=7 * side=7 * (5 * num=2 + classes=1) = 539`

Also changed `labels.txt` to contain just my 1 class label.

I started training with this command:
```
python3 flow --model cfg/bcard_tiny_yolov1.cfg --savepb --save 1000 --epoch 5000 --dataset ~/Downloads/business_cards_1st/imgs/ --annotation ~/Downloads/business_cards_1st/ann/ --threshold .01 --gpu 1.0 --lr 5e-4 --batch 64 --train --trainer adam
```

Using 23 images + VOC annotations that look like this:
```
<annotation verified="no">
  <folder>imgs</folder>
  <filename>b0.jpg</filename>
  <path>/home/mj/Downloads/business_cards/imgs/b0.jpg</path>
  <source>
    <database>Unknown</database>
  </source>
  <size>
    <width>640</width>
    <height>480</height>
    <depth>3</depth>
  </size>
  <segmented>0</segmented>
  <object>
    <name>businesscard</name>
    <pose>Unspecified</pose>
    <truncated>0</truncated>
    <difficult>0</difficult>
    <bndbox>
      <xmin>243</xmin>
      <ymin>38</ymin>
      <xmax>543</xmax>
      <ymax>340</ymax>
    </bndbox>
  </object>
</annotation>
```

After letting the loss reach ~0.05, I ran the checkpoint against some test images. Every single image returned the exact same set of 4 bounding boxes, but none of them were accurate:
```
python3 flow --model cfg/bcard_tiny_yolov1.cfg --load -1 --threshold .01 --json --test test/
```
```
[{"label": "businesscard","confidence": 0.07,"topleft": {"x": 157, "y": 74},"bottomright": {"x": 321,"y": 413}}, 
{"label": "businesscard","confidence": 0.24,"topleft": {"x": 193, "y": 171},"bottomright": {"x": 393,"y": 482}}, 
{"label": "businesscard","confidence": 0.13,"topleft": {"x": 61, "y": 215},"bottomright": {"x": 292,"y": 534}}, 
{"label": "businesscard","confidence": 0.01,"topleft": {"x": 274, "y": 363},"bottomright": {"x": 274,"y": 363}}]
```

Any help on why this is happening would be much appreciated!

Here's what the bounding boxes look like on the image (b0.jpg):
<img width="641" alt="screen shot 2017-04-18 at 6 06 27 pm" src="https://cloud.githubusercontent.com/assets/466968/25159069/d41123f6-2461-11e7-855e-86cd49953023.png">

**Comment by mjohnst** [comment]
I also tried this with a smaller training set (just 6 images). I let run until loss wasn't moving (around .05). Ran test on exact same images, and got same bounding boxes problem (all same and incorrect).

**Comment by jcarletgo** [comment]
I've never tried training with using predefined weights from the `--load`, but I've also had issues training tiny-yolo(v2) otherwise. The probs for all the anchors boxes seem to come out really close together, but the x,y,w,h seem okay otherwise, so all 5 anchor boxes show up.

**Comment by thtrieu** [comment]
> all same and incorrect.

One most likely situation is that this indicates the weights of at least one convolution has converged to (near) zero, leaving only the bias to affect prediction, therefore all images receive the same wrong prediction.

So this might be an issue of initialization. I initialize all weights to be gaussian, a xavier replacement might solve the problem. Any suggestion is welcome!

**Comment by mjohnst** [comment]
I think the issue was related to the model cfg. I've moved to a different one and am getting more accurate results now.

**Comment by astaroth106** [comment]
I am having the same problem, would you mind sharing your cfg file?

**Comment by NadineSf** [comment]
Hello @mjohnst, I am having the same problem when training with only one class.
I am using tiny-yolo model, can you please share with me which model cfg did u end up using?
Thank you

**Comment by mjohnst** [comment]
I think this is the cfg that ended up working (1 class):
```
[net]
batch=64
subdivisions=64
height=448
width=448
channels=3
momentum=0.9
decay=0.0005

learning_rate=0.0001
policy=steps
steps=20,40,60,80,20000,30000
scales=5,5,2,2,.1,.1
max_batches = 40000

[crop]
crop_width=448
crop_height=448
flip=0
angle=0
saturation = 1.5
exposure = 1.5

[convolutional]
filters=16
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
filters=256
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
filters=512
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
filters=1024
size=3
stride=1
pad=1
activation=leaky

[convolutional]
filters=1024
size=3
stride=1
pad=1
activation=leaky

[convolutional]
filters=1024
size=3
stride=1
pad=1
activation=leaky

[connected]
output=256
activation=linear

[connected]
output=4096
activation=leaky

[dropout]
probability=.5

[connected]
output= 539
activation=linear

[detection]
classes=1
coords=4
rescore=1
side=7
num=2
softmax=0
sqrt=1
jitter=.2

object_scale=1
noobject_scale=.5
class_scale=1
coord_scale=5

```

**Comment by ambr89** [comment]
Hi! I'he a question,
you have set 

> height=448
> width=448

beacuse all of your data have size 448x448?
Can I train with data with different size?
If I can, what can I set height and width ?

**Comment by yvtheja** [comment]
@thtrieu checked with Xavier initialization. Still the boxes are same for every image. 
Any other suggestions? 
Do I need to check the testing script once?

Thanks in advance :)

**Comment by groot-1313** [comment]
Hi, 
I am facing the same problem using YOLO v1 cfg - tiny-yolo.cfg. I am training from scratch on a custom data set with 2 classes. Loss reduced to 0.9 after 720 steps. But upon testing the model I get the same predictions, same bounding box predictions for all images. Has anyone found a fix to this?

**Comment by dchatterjee172** [comment]
did anyone find any explanation for this?
